### PR TITLE
Usergroups

### LINKED ISSUE #63: Add support for usergroups methods

### PR DESCRIPTION
**Pull request by whiteley**
Trying to continue from the previous work to support usergroups

Closes #63 #76

### CODE REVIEW

File: usergroups.go [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+	Description string `json:"description"`
+	Handle      string `json:"handle"`
+	IsExternal  bool   `json:"is_external"`
+	DateCreate  int    `json:"date_create"`
+	DateUpdate  int    `json:"date_update"`
+	DateDelete  int    `json:"date_delete"`
+	AutoType    string `json:"auto_type"`
+	CreatedBy   string `json:"created_by"`
+	UpdatedBy   string `json:"updated_by"`
+	DeletedBy   string `json:"deleted_by"`
```
Comment (dlsniper) [inline usergroups.go]: According to the documentation here: https://api.slack.com/types/usergroup ` deleted_by ` can be ` null ` so I guess we need to use ` *string ` as type here.

File: usergroups.go [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+// UserGroup contains all the information of a user group
+type UserGroup struct {
+	ID          string `json:"id"`
+	TeamID      string `json:"team_id"`
+	IsUserGroup bool   `json:"is_usergroup"`
+	Name        string `json:"name"`
+	Description string `json:"description"`
+	Handle      string `json:"handle"`
+	IsExternal  bool   `json:"is_external"`
+	DateCreate  int    `json:"date_create"`
```
Comment (dlsniper) [inline usergroups.go]: This should be of type ` JSONTime `

File: usergroups.go [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+	DateUpdate  int    `json:"date_update"`
```
Comment (dlsniper) [inline usergroups.go]: This should be of type ` JSONTime `

File: usergroups.go [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+	DateDelete  int    `json:"date_delete"`
```
Comment (dlsniper) [inline usergroups.go]: This should be of type ` JSONTime `

File: usergroups.go [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+	AutoType    string `json:"auto_type"`
+	CreatedBy   string `json:"created_by"`
+	UpdatedBy   string `json:"updated_by"`
+	DeletedBy   string `json:"deleted_by"`
+	Prefs       struct {
```
Comment (dlsniper) [inline usergroups.go]: This should be extracted in its own type I think so that it's easier to reference / maintain in the future.

**Review by dlsniper — CHANGES_REQUESTED** [review]
Hi @whiteley, 

Thanks a lot for your PR.
There are a few changes that should be made before this can be merged.
I really appreciate your PR and the presence of tests.

Thank you.

File: usergroups.go (hunk 1 above)
Comment (whiteley) [reply usergroups.go]: `deleted_by` can be null or an ID but not an empty string as well, it seemed like that would only be necessary if distinguishing between all three?

File: usergroups.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+	IsExternal  bool           `json:"is_external"`
+	DateCreate  JSONTime       `json:"date_create"`
+	DateUpdate  JSONTime       `json:"date_update"`
+	DateDelete  JSONTime       `json:"date_delete"`
+	AutoType    string         `json:"auto_type"`
+	CreatedBy   string         `json:"created_by"`
+	UpdatedBy   string         `json:"updated_by"`
+	DeletedBy   string         `json:"deleted_by"`
+	Prefs       UserGroupPrefs `json:"prefs"`
+	Users       []string
```
Comment (dlsniper) [inline usergroups.go]: I'm not sure if this is needed by the looks of it as ` userGroupResponseFull ` already has another ` Users ` and it's tagged while this isn't used anywhere.

File: usergroups.go (hunk 1 above)
Comment (dlsniper) [reply usergroups.go]: As we don't need to encode those I guess it's ok, fair point.

**Review by dlsniper — CHANGES_REQUESTED** [review]
Hi @whiteley, really sorry for this one last thing, I've missed in my original review. Thank you so much for this.

**Comment by dlsniper** [comment]
Turns out I can use a cool new Github feature to help on this.
Thanks a lot for the PR @whiteley, awesome job!

**Review by dlsniper — APPROVED** [review]

**Pull request merged**